core/src/test/java/org/apache/brooklyn/core/entity/EntityConfigTest.java

```diff
@@ -244,59 +252,152 @@ public void testGetConfigMapWithSubValueAsStringNotCoerced() throws Exception {
+            assertTrue(entity.config().getNonBlocking(TestEntity.CONF_NAME).isAbsent());
+            log.trace("got absent");
+            
+            latch.release();
+            
+            // Can now finish task, so will return expectedVal
+            log.trace("get blocking");
+            assertEquals(entity.config().get(TestEntity.CONF_NAME), expectedVal);
+            log.trace("got blocking");
+            assertEquals(entity.config().getNonBlocking(TestEntity.CONF_NAME).get(), expectedVal);
+            
+            latch.acquire();
+            log.trace("finished");
+        }
         
-        assertEquals(entity.config().getNonBlocking(TestEntity.CONF_MAP_OBJ_THING).get(), ImmutableMap.of("mysub", expectedVal));
-        assertEquals(entity.config().getNonBlocking(TestEntity.CONF_MAP_OBJ_THING.subKey("mysub")).get(), expectedVal);
+        protected void runGetConfigNonBlockingInMap() throws Exception {
+            Preconditions.checkNotNull(blockingVal, "Fixture must set blocking val before running this");
+            TestEntity entity = (TestEntity) mgmt.getEntityManager().createEntity(EntitySpec.create(TestEntity.class)
+                    .configure(TestEntity.CONF_MAP_OBJ_THING, ImmutableMap.<String, Object>of("mysub", blockingVal)));
+            
+            // Will initially return absent, because task is not done
+            assertTrue(entity.config().getNonBlocking(TestEntity.CONF_MAP_OBJ_THING).isAbsent());
+            assertTrue(entity.config().getNonBlocking(TestEntity.CONF_MAP_OBJ_THING.subKey("mysub")).isAbsent());
+            
+            latch.release();
+            
+            // Can now finish task, so will return expectedVal
+            assertEquals(entity.config().get(TestEntity.CONF_MAP_OBJ_THING), ImmutableMap.of("mysub", expectedVal));
+            assertEquals(entity.config().get(TestEntity.CONF_MAP_OBJ_THING.subKey("mysub")), expectedVal);
+            
+            assertEquals(entity.config().getNonBlocking(TestEntity.CONF_MAP_OBJ_THING).get(), ImmutableMap.of("mysub", expectedVal));
+            assertEquals(entity.config().getNonBlocking(TestEntity.CONF_MAP_OBJ_THING.subKey("mysub")).get(), expectedVal);
+        }
     }
     
+    @Test(groups="Integration") // because takes 1s+
+    public void testGetTaskNonBlockingKey() throws Exception {
+        new ConfigNonBlockingFixture().usingTask().runGetConfigNonBlockingInKey(); }
+    @Test(groups="Integration") // because takes 1s+
+    public void testGetTaskNonBlockingMap() throws Exception {
+        new ConfigNonBlockingFixture().usingTask().runGetConfigNonBlockingInMap(); }
+    
+    @Test(groups="Integration") // because takes 1s+
+    public void testGetTaskFactoryNonBlockingKey() throws Exception {
+        new ConfigNonBlockingFixture().usingTaskFactory().runGetConfigNonBlockingInKey(); }
+    @Test(groups="Integration") // because takes 1s+
+    public void testGetTaskFactoryNonBlockingMap() throws Exception {
```

Comment:
Why does this now work? We still have the call to `task.cancel(true)` in `AbstractConfigurationSupportInternal.getNonBlockingResolvingStructuredKey()`, so during the first call to `getNonBlocking` we'll cancel the task. Is it because that is a different task that does not pass the `cancel()` call down to the actual config-key-value task that we are trying to evaluate?